RabbitMQ on Kubernetes/Portworx - Reference Architecture (#492)
portworx/pxdocs: content/portworx-install-with-kubernetes/application-install-with-kubernetes/rabbitmq/_index.md

1. The spec below creates a `StorageClass` with 3 replicas and uses a high-priority storage pool:

```text
kubectl apply -f - <<'_EOF'
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: portworx-rabbitmq
parameters:
  io_priority: high
  repl: "2"
  group: "rmq_vg"
provisioner: kubernetes.io/portworx-volume
allowVolumeExpansion: true
reclaimPolicy: Delete
_EOF
```

```output
storageclass.storage.k8s.io/portworx-rabbitmq created
```

For details about the Portworx-specific parameters, refer to the [Portworx Volume](https://kubernetes.io/docs/concepts/storage/storage-classes/#portworx-volume) section.

{{<info>}}
If you're using Portworx with CSI, you need to set the value of the `provisioner` parameter to `pxd.portworx.com`.
{{</info>}}

When you install RabbitMQ with Portworx on Kubernetes, you can choose one of the following options:

* Set up RabbitMQ using [Helm](https://helm.sh)]
* Set up RabbitMQ manually

Setup RabbitMQ (using Helm)

> ### Helm

Reviewer: ```suggestion
Set up RabbitMQ using Helm
```

Author: fixed

> A complex application like RabbitMQ relies on several YAML files to define its various components. The next chapter will show how you can use Helm to simplify the deployment of RabbitMQ.

Reviewer: ```suggestion
AA complex application like RabbitMQ relies on several YAML files to define its various components. This chapter shows how you can use Helm to simplify the deployment of RabbitMQ.
```

Author: fixed tense, but used the word `section` instead of `chapter`, if that's ok?

The following Helm command uses the [RabbitMQ High Available](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha) Helm chart to create a [release](https://github.com/helm/helm/blob/release-2.14/docs/glossary.md#release) named rmq. The release creates a two-member RabbitMQ cluster, and Portworx will mirror queues and messages between the nodes.

```text
helm upgrade \
  --install \
  --set replicaCount=2 \
  --set rabbitmqUsername=admin \
  --set rabbitmqPassword=secretpassword \
  --set managementPassword=anothersecretpassword \
  --set rabbitmqErlangCookie=secretcookie  \
  --set persistentVolume.enabled=true \
  --set persistentVolume.storageClass=portworx-rabbitmq \
  --set schedulerName=stork \
rmq stable/rabbitmq-ha
```

..to which, after a minute or two, you should get back the a response from Helm confirming success and specifics of the release we created (including details of how it is configured/accessed):

```output
Release "rmq" does not exist. Installing it now.

NAME: rmq
LAST DEPLOYED: Mon Oct 14 16:58:20 2019
NAMESPACE: default
STATUS: DEPLOYED
```

> Other than specifying that this should be a two-replica [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), it is worth mentioning that it is _here_ we referenced that StorageClass we defined above, and we also have specify a few dummy credentials (which are insecure and should _not_ be used in anything but a testing environment).

Reviewer: ```suggestion
Note that the `replicaCount` variable in this example creates a two-replica [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) and the `persistentVolume.storageClass` references the `StorageClass` defined above.
```

Author: fixed

> Alternatively, the following section describes how to manually do what helm did for us above.

Reviewer: We can probably remove this.

> ## Setup RabbitMQ manually

Reviewer: ```suggestion
Set up RabbitMQ manually
```

Author: fixed

> In this section we will instead take the classic approach of including here all the various yaml definitions needed to set up RabbitMQ.

Reviewer: @aleks-mariusz Why would a user want to install RabbitMQ this way? Probably we should elaborate a bit.

Author: well they may not all have helm or want to install helm, so i provide an alternative way.. shall i mention exactly that?

Reviewer: @aleks-mariusz Hmmm... I don't think so. I was trying to understand if there are situations in which a user would prefer to set up RabbitMQ manually.  This is too obvious, I wouldn't mention it. Thanks!

{{<info>}}

> The contents of the yaml used below, are largely based on what the [RabbitMQ chart](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha) [templates](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha/templates) create when using Helm.

Reviewer: ```suggestion
Note that the spec files used in this chapter are based on these [Helm templates](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha/templates).
```

Author: fixed

{{</info>}}

### Configuration and credentials

> First we will set-up a [ConfigMap and Secret](https://kubernetes.io/blog/2016/04/configuration-management-with-containers/) that will configure how RabbitMQ starts up and "secure" it with the same dummy credentials.

Reviewer: ```suggestion
Enter the following command to create a `ConfigMap` and a Kubernetes secret:
```

Author: fixed

> ```text

Reviewer: @nathan-px This listing is pretty long. I'd just point the users to a link where they can download the spec and apply it.

  labels:
    app: rabbitmq-ha
type: Opaque
data:
  rabbitmq-username: "YWRtaW4="
  rabbitmq-password: "c2VjcmV0cGFzc3dvcmQ="
  rabbitmq-management-username: "bWFuYWdlbWVudA=="
  rabbitmq-management-password: "YW5vdGhlcnNlY3JldHBhc3N3b3Jk"
  rabbitmq-erlang-cookie: "c2VjcmV0Y29va2ll"
  definitions.json: "ewogICJnbG9iYWxfcGFyYW1ldGVycyI6IFsKICAgIAogIF0sCiAgInVzZXJzIjogWwogICAgewogICAgICAibmFtZSI6ICJtYW5hZ2VtZW50IiwKICAgICAgInBhc3N3b3JkIjogImFub3RoZXJzZWNyZXRwYXNzd29yZCIsCiAgICAgICJ0YWdzIjogIm1hbmFnZW1lbnQiCiAgICB9LAogICAgewogICAgICAibmFtZSI6ICJhZG1pbiIsCiAgICAgICJwYXNzd29yZCI6ICJzZWNyZXRwYXNzd29yZCIsCiAgICAgICJ0YWdzIjogImFkbWluaXN0cmF0b3IiCiAgICB9CiAgXSwKICAidmhvc3RzIjogWwogICAgewogICAgICAibmFtZSI6ICIvIgogICAgfQogIF0sCiAgInBlcm1pc3Npb25zIjogWwogICAgewogICAgICAidXNlciI6ICJhZG1pbiIsCiAgICAgICJ2aG9zdCI6ICIvIiwKICAgICAgImNvbmZpZ3VyZSI6ICIuKiIsCiAgICAgICJyZWFkIjogIi4qIiwKICAgICAgIndyaXRlIjogIi4qIgogICAgfQogIF0sCiAgInBhcmFtZXRlcnMiOiBbCiAgICAKICBdLAogICJwb2xpY2llcyI6IFsKICAgIAogIF0sCiAgInF1ZXVlcyI6IFsKICAgIAogIF0sCiAgImV4Y2hhbmdlcyI6IFsKICAgIAogIF0sCiAgImJpbmRpbmdzIjogWwogICAgCiAgXQp9"
_EOF
```

```output
configmap/rmq-rabbitmq-ha created
secret/rmq-rabbitmq-ha created
```

{{<info>}}

> The above credentials are the same dummy (insecure) credentials as mentioned above in the previous approach that utilized helm.

Reviewer: ```suggestion
Note that we're separating the application code from the configuration. Please refer to the [Configuration management with Containers](https://kubernetes.io/blog/2016/04/configuration-management-with-containers/) for more details about the `ConfigMap` resource.
```

Author: fixed

{{</info>}}

> ### RBAC for RabbitMQ

Reviewer: ```suggestion
Set up RBAC
```

Author: fixed

> This section will set up RBAC-related objects for the workloads we'll define in the [next section](#rabbitmq-cluster).

Author: fixed (w/o typo)

```text
kubectl apply -f - <<'_EOF'
---
# Source: rabbitmq-ha/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app: rabbitmq-ha
  name: rmq-rabbitmq-ha
  namespace: default
---
# Source: rabbitmq-ha/templates/role.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  labels:
    app: rabbitmq-ha
  name: rmq-rabbitmq-ha
  namespace: default
rules:
  - apiGroups: [""]
    resources: ["endpoints"]
    verbs: ["get"]
---
# Source: rabbitmq-ha/templates/rolebinding.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  labels:
    app: rabbitmq-ha
  name: rmq-rabbitmq-ha
  namespace: default
subjects:
  - kind: ServiceAccount
    name: rmq-rabbitmq-ha
    namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: rmq-rabbitmq-ha
_EOF
```

```output
serviceaccount/rmq-rabbitmq-ha created
role.rbac.authorization.k8s.io/rmq-rabbitmq-ha created
rolebinding.rbac.authorization.k8s.io/rmq-rabbitmq-ha created
```

### RabbitMQ cluster

> Here we're finally launching the workload definition, which consists of the StatefulSet and some supporting Services.

Reviewer: ```suggestion
Use the following command to create a `StatefulSet` and the required supporting services:
```

Author: fixed

> ```text

Reviewer: @nathan-px This listing is pretty long. I'd just point the users to a link where they can download the spec and apply it.

> ```text

Reviewer: @nathan-px Again, I'd just point the users to a link where they can download the spec and apply it.

Author: we should standardize on this approach, i don't like having half the yaml in-line in the document and another half being linked/needed to be downloaded.. IMO it's a good practice to be able to read/examine it instead of just blindly downloading things and running them (at least from a security perspect), inlineing things, even though they may be long, makes that an easier task (vs having to download, open in an editor, switch back to the doc, etc).

Reviewer: @aleks-mariusz Yeah, I see your point. Maybe we could fold these long snippets by default. Anyway, this is a larger problem. We don't need to fix it as part of this PR :)

Author: folding sounds good, is this possible today (meaning there a differnt tag i should be using) or is this a future-enhancement where it can remain this way for now ?

            - weight: 1
              podAffinityTerm:
                topologyKey: "kubernetes.io/hostname"
                labelSelector:
                  matchLabels:
                    app: rabbitmq-ha
      volumes:
        - name: config
          emptyDir: {}
        - name: configmap
          configMap:
            name: rmq-rabbitmq-ha
        - name: definitions
          secret:
            secretName: rmq-rabbitmq-ha
            items:
            - key: definitions.json
              path: definitions.json
  volumeClaimTemplates:
    - metadata:
        name: data
        annotations:
      spec:
        accessModes:
          - "ReadWriteOnce"
        resources:
          requests:
            storage: "8Gi"
        storageClassName: "portworx-rabbitmq"
_EOF
```

```output
service/rmq-rabbitmq-ha-discovery created
service/rmq-rabbitmq-ha created
statefulset.apps/rmq-rabbitmq-ha created
```

Note that the last line of the spec references the `portworx-rabbitmq` storage class defined in the [Create a StorageClass](#create-a-storage-class) section. As a result, Kubernetes will automatically create a new PVC for each replica.

## Validate the cluster functionality

> Regardless of the method we used to setup RabbitMQ in Kubernetes, we should now be able to control and test our RabbitMQ cluster.

Reviewer: @aleks-mariusz I'd rephrase these two paragraphs like this:

```
Regardless of the method you used to set up RabbitMQ, you can test your installation using a tool called [PerfTest](https://rabbitmq.github.io/rabbitmq-perf-test/stable/htmlsingle/). In the next sections, we will use PerfTest to:

* simulate basic workloads
* measure the performance of the system
* introduce a failure and see how the cluster recovers from it.
```

Is this accurate?

> You may have spotted that the last line reference that StorageClass created near the beginning of this document.  As a result of the `volumeClaimTemplate` section, a new PersistentVolumeClaim will also be created automatically by Kubernetes for each replica.

Reviewer: ```suggestion
Note that the last line of the spec references the `portworx-rabbitmq` storage class defined in the [Create a StorageClass](#create-a-storage-class) section. As a result, Kubernetes will automatically create a new PVC for each replica.
```

Author: fixed

> ## Post-install validation testing

Reviewer: ```suggestion
Validate the cluster functionality
```

Author: fixed, but with all these types of suggestions, we're effectively removing the ability to move around the document easily as each of these section headers show up as separate TOC entries (that make navigating around the doc possible)

We will use [PerfTest](https://rabbitmq.github.io/rabbitmq-perf-test/stable/htmlsingle/), a testing suite bundled with RabbitMQ, to simulate broker use, measure performance of the system and perform failover testing.

### Example RabbitMQ policy for H/A

> First we will configure a queue-policy named `perf-test-with-ha` to:

Reviewer: ```suggestion
First, we configure a policy named `perf-test-with-ha`  with the following settings:

* It's only applied on queues that begin with `perf-test` <!--revisit this -->
* Sets the [ha-mode parameter].(https://www.rabbitmq.com/ha.html#mirroring-arguments) equal to `2`.
* Places the [queue-masters](https://www.rabbitmq.com/ha.html#behaviour) on the least-loaded node.
* Sets up the queues as [lazy](https://www.rabbitmq.com/lazy-queues.html). RabbitMQ will save the queues to disk as early as possible.
```

  * match only queues that begin with `perf-test`
  * set the [ha-mode parameter](https://www.rabbitmq.com/ha.html#mirroring-arguments) to be exactly 2
  * locate the [queue-masters](https://www.rabbitmq.com/ha.html#behaviour) on whichever is the least loaded node
  * finally set up these queues to be [_lazy_](https://www.rabbitmq.com/lazy-queues.html), which will make them want to use storage for persistance early/automatically

```text
kubectl exec rmq-rabbitmq-ha-0 -- \
  rabbitmqctl set_policy perf-test-with-ha '^perf-test' \
'{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}' --apply-to queues
```

```output

> Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{

Reviewer: ```suggestion
Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}" with priority "0" for vhost "/" ...
```

> Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{

Reviewer: The output should be:

```
Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}" with priority "0" for vhost "/" ...
```

Author: do you mean we should not truncate it? I only avoided repeating it because it's just a confirmation of the exact params that were just given (so i included output more for the sake of knowing this step was successful; that is to say, if it wasn't, it wouldn't even say `setting policy..`

Reviewer: For consistency, I'd either drop the output or include it in full.

...
}" with priority "0" for vhost "/" ...
```

### Monitoring

> RabbitMQ has a wonderful enabled-by-default [plugin for a web-UI](https://www.rabbitmq.com/management.html) that can be accessed by port-forwarding port 15672 to your local system.  It is useful to have this loaded during the testing to monitor the system.

Reviewer: ```suggestion
RabbitMQ provides a web-based [user interface](https://www.rabbitmq.com/management.html) that you can use to manage your cluster. To access it, you must forward connections made to a local port to the 15672 port on the   `rmq-rabbitmq-ha-0` pod.

Enter the following `kubectl port-forward` command to forward all connections made to `localhost:15672` to `rmq-rabbitmq-ha-0:15672`:
```

Author: fixed

      ```

      ```output
      Forwarding from 127.0.0.1:15672 -> 15672
      ```

> 2. You can now access the RabbitMQ user interface by pointing your browser to [localhost:1567](localhost:1567). Use the credentials created in the [Set up RabbitMQ using Helm](#set-up-rabbitmq-using-helm) section.

Reviewer 2: ```suggestion
2. You can now access the RabbitMQ user interface by pointing your browser to `localhost:1567`. Use the credentials created in the [Set up RabbitMQ using Helm](#set-up-rabbitmq-using-helm) section.
```

> 3. Visit the [Queue section](http://127.0.0.1:15672/#/queues), to see that the `perf-test-with-ha` policy is applied.

Reviewer 2: These internal IPs don't validate. They may work for our users, but our docs won't build with them. 

```suggestion
3. Visit the [Queue section](http://127.0.0.1:15672/#/queues), to see that the `perf-test-with-ha` policy is applied.
```

            - weight: 1
              podAffinityTerm:
                topologyKey: "kubernetes.io/hostname"
                labelSelector:
                  matchLabels:
                    app: rabbitmq-ha
      volumes:
        - name: config
          emptyDir: {}
        - name: configmap
          configMap:
            name: rmq-rabbitmq-ha
        - name: definitions
          secret:
            secretName: rmq-rabbitmq-ha
            items:
            - key: definitions.json
              path: definitions.json
  volumeClaimTemplates:
    - metadata:
        name: data
        annotations:
      spec:
        accessModes:
          - "ReadWriteOnce"
        resources:
          requests:
            storage: "8Gi"
        storageClassName: "portworx-rabbitmq"
_EOF
```

```output
service/rmq-rabbitmq-ha-discovery created
service/rmq-rabbitmq-ha created
statefulset.apps/rmq-rabbitmq-ha created
```

Note that the last line of the spec references the `portworx-rabbitmq` storage class defined in the [Create a StorageClass](#create-a-storage-class) section. As a result, Kubernetes will automatically create a new PVC for each replica.

## Validate the cluster functionality

> Regardless of the method we used to setup RabbitMQ in Kubernetes, we should now be able to control and test our RabbitMQ cluster.

Reviewer: ```suggestion
Regardless of the method you used to set up RabbitMQ, you can test your installation using a tool called [PerfTest](https://rabbitmq.github.io/rabbitmq-perf-test/stable/htmlsingle/). In the next sections, we will use PerfTest to:

* simulate basic workloads
* measure the performance of the system
* introduce a failure and see how the cluster recovers from it.
```

Author: fixed

* Places the [queue-masters](https://www.rabbitmq.com/ha.html#behaviour) on the least-loaded node.
* Sets up the queues as [lazy](https://www.rabbitmq.com/lazy-queues.html). RabbitMQ will save the queues to disk as early as possible.

```text
kubectl exec rmq-rabbitmq-ha-0 -- \
  rabbitmqctl set_policy perf-test-with-ha '^perf-test' \
'{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}' --apply-to queues
```

```output
Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{
...
}" with priority "0" for vhost "/" ...
```

### Monitoring

RabbitMQ provides a web-based [user interface](https://www.rabbitmq.com/management.html) that you can use to manage your cluster. To access it, you must forward connections made to a local port to the 15672 port on the   `rmq-rabbitmq-ha-0` pod.

Enter the following `kubectl port-forward` command to forward all connections made to `localhost:15672` to `rmq-rabbitmq-ha-0:15672`:

```text
kubectl port-forward rmq-rabbitmq-ha-0 15672:15672
```

```output
Forwarding from 127.0.0.1:15672 -> 15672
```

> At this point you should be able to hit TCP port 15672 on [_localhost_](http://127.0.0.1:15672) in your web browser, and see the RabbitMQ WebUI.  You can use the credentials referenced [earlier](#launching-rabbitmq-release) in this document to log in, and explore.   If you then visit the [Queue section](http://127.0.0.1:15672/#/queues),  once the performance test starts and the queues are created, you should see the policy we setup be applied as well.

Reviewer: @aleks-mariusz What if the user chooses to set up RabbitMQ manually? It looks like we define these credentials only if the user chooses to install RabbitMQ using Helm. Probably I'm missing something.

Author: no, these same credentials are encoded/included in the secrets we created when doing it the _classic_ way.. this was made a note of in the docs but i was asked to remove it :-P

Reviewer: Also, I'd rewrite this as follows:

```
At this point, you can access the RabbitMQ user interface by pointing your browser to [localhost:1567](localhost:1567). Use the credentials created in the [#set-up-rabbitmq-using-helm].

Once you start the performance tests, you can then visit the [Queue section](http://127.0.0.1:15672/#/queues), to see that the `perf-test-with-ha` policy is applied.
```

Reviewer: @aleks-mariusz Gotcha. Thanks!

> At this point you should be able to hit TCP port 15672 on [_localhost_](http://127.0.0.1:15672) in your web browser, and see the RabbitMQ WebUI.  You can use the credentials referenced [earlier](#launching-rabbitmq-release) in this document to log in, and explore.   If you then visit the [Queue section](http://127.0.0.1:15672/#/queues),  once the performance test starts and the queues are created, you should see the policy we setup be applied as well.

Reviewer: ```suggestion
At this point, you can access the RabbitMQ user interface by pointing your browser to [localhost:1567](localhost:1567). Use the credentials created in the [#set-up-rabbitmq-using-helm].

Once you start the performance tests, you can then visit the [Queue section](http://127.0.0.1:15672/#/queues), to see that the `perf-test-with-ha` policy is applied.
```

Author: fixed

> Additionally, one can examine the log-messages of the pod that will survive the failover test:

Reviewer: ```suggestion
Additionally, you can stream to `stdout` the logs of the pod that survives the failover test by entering the following command:
```

Author: fixed

```text
kubectl logs -f rmq-rabbitmq-ha-0
```

> This will show (after the pod has started) the log-messages from the container live.

Reviewer: It seems redundant.

Author: agreed, removed.. (i only included this to be consistent with every other command describing/including the output of the previous command)

### Containerized testing environment

> Here we spawn a pod with one container which will allow us to run PerfTest in the [_next_ step](#simulate-broker-usage):

Reviewer: ```suggestion
The following `kubectl run` command creates a POD with a sleeping container on a node that doesn't run RabbitMQ. The container runs for an hour (see the `"args": ["-c", "sleep 3600"]` line). Then, Kubernetes automatically restarts the container (see the `--restart=Always` line).
```

```text
kubectl run perftest \
  --restart=Always \
  --image dummy-required-param \
  --generator=run-pod/v1 \
  --overrides \
'{
  "spec": {
    "affinity": {
      "podAntiAffinity": {
        "requiredDuringSchedulingIgnoredDuringExecution": [{
          "labelSelector": {
            "matchExpressions": [{
              "key": "app",
              "operator": "In",
              "values": ["rabbitmq-ha"]
            }]
          },
          "topologyKey": "kubernetes.io/hostname"
        }]
      }
    },
    "containers": [{
      "command": ["sh"],
      "image": "pivotalrabbitmq/perf-test:latest",
      "name": "perftest",
      "args": ["-c", "sleep 3600"]
    }],
    "securityContext": {
      "runAsUser": 1000
    }
  }
}'
```

```output
pod/perftest created
```

> A pod with a sleeping container is started on one of the _other_ nodes (that _aren't_ running RabbitMQ).  This container runs for an hour (by sleeping 3600 seconds) and always restart when finished.   We also specify a non-root user to run this environment as, since your cluster may have PodSecurityPolicies enabled.  In the next step we will `kubectl exec` into this pod.

Reviewer: ```suggestion
{{<info>}}
Note that we use a non-root user to run the commands. This is useful for when you have `PodSecurityPolicies` enabled.
{{</info>}}
```

Author: i reworded this in c9e5400 in my own way, can you take a look and let me know if this is acceptable?

> ### Simulate Broker Usage

Reviewer: ```suggestion
### Simulate basic workloads
```

Author: fixed

> Now we're ready to start up the simulated producers/consumers inside the pre-existing pod's container from the [previous step](#containerized-testing-environment).

Reviewer: ```suggestion
To start the simulated producers and consumers, run the `bin/runjava` command in the `perftest` container:
```

Author: fixed

```text
kubectl exec -it perftest -- \
    bin/runjava com.rabbitmq.perf.PerfTest \
      --time 900 \
      --queue-pattern 'perf-test-%d' \
      --queue-pattern-from 1 \
      --queue-pattern-to 2 \
      --producers 2 \
      --consumers 8 \
      --queue-args x-cancel-on-ha-failover=true \
      --flag persistent \
      --uri amqp://admin:secretpassword@rmq-rabbitmq-ha:5672?failover=failover_exchange
```

> ..to which you should get back the output of the utility itself.

Reviewer: Please delete this.

Author: removed

> The test-suite's parameters themselves do the following:

Reviewer: ```suggestion
There are several things to note about this example:

* Each session runs for 15 minutes, and the pattern of the queue's name matches `perf-test-%d` followed by `1` or `2`.
* We create 2 threads for producers and 8 threads for consumers. Thus, each RabbitMQ node should host 1 queue with 1 producer and 4 consumers.
* We flag the messages as persistent so that RabbitMQ saves them to disk.
```

Author: fixed

* run the session for 15 minutes with queues named in the format of `perf-test-` followed by a number (in our case `1` and `2`)
* create threads for 2 producers and 8 consumers (so theoretically each of the two rabbitmq replicas should host 1 queue and get 1 producers and 4 consumers
* set messages to be flagged as persistent (so will always involve IO which should be understood that this trades off performance for reliabilty)
* use the queue with the sample credentials, at the expected address from the previously defined service (and is internal to the kubernetes cluster)

### Failover

> Here we can create a scenario for failure, such as killing one of the pods.

Reviewer: ```suggestion
Simulate a failure scenario by deleting the `rmq-rabbitmq-ha-1` pod:
```

Author: fixed

```text
kubectl delete pod rmq-rabbitmq-ha-1 --force --grace-period=0
```

```output
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
pod "rmq-rabbitmq-ha-1" force deleted
```

> The performance testing session may pause momentarily (and messages in the queue pile up) as the system recovers from the failure, the queue fails-over eventualy resume with no messages being lost (presuming enough time remains during testing, for all the the testing consumers to catch up).

Reviewer: ```suggestion
As the system recovers from the failure, you will see that the performance testing session will pause for a short period. As a result,  the messages will pile up in the queue. After the failover, the cluster resumes normal processing, and no messages are lost. 
```

Author: fixed

## Cleaning up

> If helm was used, cleaning is as easy as first running:

Reviewer: ```suggestion
If you used Helm to install the RabbitMQ cluster, then enter the following command to remove all the Kubernetes components and delete the release:
```

Author: fixed

```text
helm delete --purge rmq
```

> Otherwise we need to manually delete all the created created resources from our cluster, by running:

Reviewer: ```suggestion
If you installed the RabbitMQ cluster manually, then enter the following command to delete all the associated resources:
```

Author: fixed

```text
kubectl delete sts/rmq-rabbitmq-ha \
  svc/rmq-rabbitmq-ha svc/rmq-rabbitmq-ha-discovery \
  rolebinding/rmq-rabbitmq-ha \
  role/rmq-rabbitmq-ha \
  sa/rmq-rabbitmq-ha \
  secret/rmq-rabbitmq-ha \
  cm/rmq-rabbitmq-ha
```

> Finally (regardless of deployment method), we throw away the data-volumes, volume-param definition and perftest pod by running:

Reviewer: @aleks-mariusz Do we need two separate `kubectl delete` commands?

Author: if we combined both kubectl commands into one, the user would see a lot of errors if they used helm as helm did the deletion already and the user would be trying to delete the (now) non-existent resources.. 

additionally, i like separating out the data destroying method and explicitly mentioning how destructive that command is so there's no surprises (not that any user should ever be using a guide like this as-is for prod purposes but i bet not outside the realm of possibilities).

```text
kubectl delete pvc/data-rmq-rabbitmq-ha-0 pvc/data-rmq-rabbitmq-ha-1 \
  sc/portworx-rabbitmq \
  po/perftest
```

The above deletes the...

  * workload itself (statefulset/services)
  * RBAC for the workload (rolebinding/role/serviceaccount)
  * configuration (configmap/secret)
  * storage volumes including data (persistentvolumeclaims)
  * volume parameters (storageclass)
  * pod we simulated broker use from (pod)

At this point you could, if you wanted to, start all over from the beginning of this document.

> ## Summary

Reviewer: @nathan-px What do you think... should we keep this section or not?

```text
kubectl apply -f - <<'_EOF'
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: portworx-rabbitmq
parameters:
  io_priority: high
  repl: "2"
  group: "rmq_vg"
provisioner: kubernetes.io/portworx-volume
allowVolumeExpansion: true
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
_EOF
```

```output
...
storageclass.storage.k8s.io/portworx-rabbitmq created
```

For details about the Portworx-specific parameters, refer to the [Powrtorx Volume]https://kubernetes.io/docs/concepts/storage/storage-classes/#portworx-volume] section.

{{<info>}}
If you're using Portworx with CSI, you need to set the value of the `provisioner` parameter to `pxd.portworx.com`.
{{</info>}}

When you install RabbitMQ with Portworx on Kubernetes, you can choose one of the following options:
* Set up RabbitMQ using [Helm](https://helm.sh)]
* Set up RabbitMQ manually

Setup RabbitMQ (using Helm)

### Helm

A complex application like RabbitMQ relies on several YAML files to define its various components. The next chapter will show how you can use Helm to simplify the deployment of RabbitMQ.

> 

Reviewer: The following Helm command uses the [RabbitMQ High Available](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha) Helm chart to create a [release](https://github.com/helm/helm/blob/release-2.14/docs/glossary.md#release) named `rmq`. This will create a two-member RabbitMQ cluster, and Portworx will mirror queues and messages between the nodes.

Author: restored in 1baa8e2

Reviewer: Thanks! 👍

```text
helm upgrade \
  --install \
  --set replicaCount=2 \
  --set rabbitmqUsername=admin \
  --set rabbitmqPassword=secretpassword \
  --set managementPassword=anothersecretpassword \
  --set rabbitmqErlangCookie=secretcookie  \
  --set persistentVolume.enabled=true \
  --set persistentVolume.storageClass=portworx-rabbitmq \
  --set schedulerName=stork \
rmq stable/rabbitmq-ha
```

> ..to which, after a minute or two, you should get back the a response from Helm confirming success and specifics of the release we created (including details of how it is configured/accessed):

Reviewer: @aleks-mariusz You mentioned that the output provides details about how the release can be accessed, but I can't spot this. Is the output truncated? If so, should we drop this or show the full output? I'm asking because I'm not sure if the users will know how to access the release.

Author: The output from Helm is pretty standard and is based on what [the chart's `./template/NOTES.txt`](https://github.com/helm/charts/blob/master/stable/rabbitmq-ha/templates/NOTES.txt) has.   

However, we kind of guide them how to access the release elsewhere in the document (since with this ref-arch we're providing a testing harness and validation steps of the cluster).

Reviewer: This is a bit confusing. Probably it would be better to provide the full output and explain how to interpret it (if needed).

Reviewer: And if we decide to keep sentence, we should move it below the output block.

Author: I think i only really referenced the output providing details so the user would know whether their helm command succeeded or not.. that it provides detials is probably not important so can be removed if you want?

```text
kubectl apply -f - <<'_EOF'
---
# Source: rabbitmq-ha/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app: rabbitmq-ha
  name: rmq-rabbitmq-ha
  namespace: default
---
# Source: rabbitmq-ha/templates/role.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  labels:
    app: rabbitmq-ha
  name: rmq-rabbitmq-ha
  namespace: default
rules:
  - apiGroups: [""]
    resources: ["endpoints"]
    verbs: ["get"]
---
# Source: rabbitmq-ha/templates/rolebinding.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  labels:
    app: rabbitmq-ha
  name: rmq-rabbitmq-ha
  namespace: default
subjects:
  - kind: ServiceAccount
    name: rmq-rabbitmq-ha
    namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: rmq-rabbitmq-ha
_EOF
```

```output
...
serviceaccount/rmq-rabbitmq-ha created
role.rbac.authorization.k8s.io/rmq-rabbitmq-ha created
rolebinding.rbac.authorization.k8s.io/rmq-rabbitmq-ha created
```

### RabbitMQ cluster

Here we're finally launching the workload definition, which consists of the StatefulSet and some supporting Services.

```text

> kubectl apply -f - <<'_EOF'

Reviewer: @aleks-mariusz This errors out:

```
Error from server (BadRequest): pod rmq-rabbitmq-ha-0 does not have a host assigned
```

The pod is stuck in `Pending`:

```
kubectl get pods -o wide -n default
NAME                READY   STATUS    RESTARTS   AGE    IP       NODE     NOMINATED NODE   READINESS GATES
rmq-rabbitmq-ha-0   0/1     Pending   0          11m   <none>   <none>   <none>           <none>
```

Author: good spot there.. we may want to tell helm to `--wait` before returning, running the testing/workload pod too quickly may result in this as the rabbitmq pods have not yet been scheduled..

Reviewer: @aleks-mariusz I ran this on our test cluster and the POD is still stuck in `Pending`:

```
kubectl get pods -o wide -n default
NAME                READY   STATUS    RESTARTS   AGE   IP       NODE     NOMINATED NODE   READINESS GATES
rmq-rabbitmq-ha-0   0/1     Pending   0          23h   <none>   <none>   <none>           <none>
```

Thoughts?

Author: it looks like it's having trouble scheduling the pod, can you use the `describe` on `rmq-rabbitmq-ha-0` instead of `get` ? does your test cluster have portworx and the storageclass was properly defined earlier ?

We will use [PerfTest](https://rabbitmq.github.io/rabbitmq-perf-test/stable/htmlsingle/), a testing suite bundled with RabbitMQ, to simulate broker use, measure performance of the system and perform failover testing.

### Example RabbitMQ policy for H/A

First we will configure a queue-policy named `perf-test-with-ha` to:

  * match only queues that begin with `perf-test`
  * set the [ha-mode parameter](https://www.rabbitmq.com/ha.html#mirroring-arguments) to be exactly 2
  * locate the [queue-masters](https://www.rabbitmq.com/ha.html#behaviour) on whichever is the least loaded node
  * finally set up these queues to be [_lazy_](https://www.rabbitmq.com/lazy-queues.html), which will make them want to use storage for persistance early/automatically

```text
kubectl exec rmq-rabbitmq-ha-0 -- \
  rabbitmqctl set_policy perf-test-with-ha '^perf-test' \
'{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}' --apply-to queues
```

```output
Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{

> ...

Reviewer: @aleks-mariusz Is this output truncated?

Author: yes, it just echo's back whatever was passed to it..

> }" with priority "0" for vhost "/" ...

Reviewer: @aleks-mariusz ditto

---
title: RabbitMQ with Portworx on Kubernetes
linkTitle: RabbitMQ
keywords: portworx, container, Kubernetes, rabbitMQ, storage, Docker, k8s, pvc
description: See how RabbitMQ can be deployed on Kubernetes using Portworx volumes.
weight: 2
noicon: true
---

[RabbitMQ](https://www.rabbitmq.com/) is an open-source message broker.  It plays a central role in many distributed systems.  You can configure RabbitMQ to safely pass messages between decoupled systems.

This reference architecture document shows how you can create and run RabbitMQ with Portworx on Kubernetes. This way, Portworx will provide a reliable persistent storage layer, which makes sure no messages are lost.

The RabbitMQ cluster will use [mirrored queues](https://www.rabbitmq.com/ha.html) to persist data and metadata to Portworx volumes. Portworx also replicates the volumes, providing multiple layers of redundancy.

## Prerequisites

* You must have a Kubernetes cluster with a minimum of 3 worker nodes.

> * Portworx is installed on your Kubernetes cluster. For more details on how to install Portworx, refer to the instructions from the [Portworx on Kubernetes](/portworx-install-with-kubernetes/) page.

Reviewer 2: ```suggestion
* Portworx must be installed on your Kubernetes cluster. For more details on how to install Portworx, refer to the instructions from the [Portworx on Kubernetes](/portworx-install-with-kubernetes/) page.
```

Author: fixed

> * (Optional) Helm in installed on your Kubernetes cluster. Refer to the [Helm Quickstart Guide](https://helm.sh/docs/using_helm/#quickstart) for details about how to install Helm.

Author: fixed in 11c7b12

## Create a StorageClass

Enter the following `kubectl apply` command to create a `StorageClass` with the following attributes:

> * The `provisioner` parameter is set to  `kubernetes.io/portworx-volume`. For details about the Portworx-specific parameters, refer to the [Portworx Volume](https://kubernetes.io/docs/concepts/storage/storage-classes/#portworx-volume) section.

Reviewer: @nathan-px Depends on the audience.I was about to remove them myself but ended up keeping them.

> * The `provisioner` parameter is set to  `kubernetes.io/portworx-volume`. For details about the Portworx-specific parameters, refer to the [Portworx Volume](https://kubernetes.io/docs/concepts/storage/storage-classes/#portworx-volume) section.

Reviewer: I moved this below the spec.

The following Helm command uses the [RabbitMQ High Available](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha) Helm chart to create a [release](https://github.com/helm/helm/blob/release-2.14/docs/glossary.md#release) named rmq. The release creates a two-member RabbitMQ cluster, and Portworx will mirror queues and messages between the nodes.

```text
helm upgrade \
  --install \
  --set replicaCount=2 \
  --set rabbitmqUsername=admin \
  --set rabbitmqPassword=secretpassword \
  --set managementPassword=anothersecretpassword \
  --set rabbitmqErlangCookie=secretcookie  \
  --set persistentVolume.enabled=true \
  --set persistentVolume.storageClass=portworx-rabbitmq \
  --set schedulerName=stork \
rmq stable/rabbitmq-ha
```

```output
Release "rmq" does not exist. Installing it now.

NAME: rmq
LAST DEPLOYED: Mon Oct 14 16:58:20 2019
NAMESPACE: default
STATUS: DEPLOYED
```

Note that the `replicaCount` variable in this example creates a two-replica [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) and the `persistentVolume.storageClass` references the `StorageClass` defined above.

## Set up RabbitMQ manually

> In this section we will instead take the classic approach of including here all the various yaml definitions needed to set up RabbitMQ.

Reviewer: ```suggestion
In this section, we'll show how you can install RabbitMQ manually.
```

Author: fixed

{{<info>}}
Note that the spec files used in this chapter are based on these [Helm templates](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha/templates).
{{</info>}}

> ### Configuration and credentials

Author: I don't follow, can you elaborate on what you mean?  The heading of "Configuration and credentials" ? I was splitting up the steps, the docs setup shows each item listed as separate items in the TOC section (the small far-right list) if they are separated out..

> ### Configuration and credentials

Reviewer 2: We typically want to split task topics by headings if the sections can be performed independently. Since these steps aren't introducing any new concept information or context, and they aren't composed of multiple substeps (one of the reasons to break up sequential instructions), I think we could combine these into a task list.

The numbering makes it easier for users to scan, recognize the tasks as a list of operations they're supposed to perform, and keep their place within the steps.

> ### Configuration and credentials

Author: ahh i see, is there an example of what a task list should look like? then i can migrate these over to that format, as per your guidance that _would_ make more sense..

> ### Configuration and credentials

Reviewer: @aleks-mariusz It's just a numbered list.

> ### Configuration and credentials

Author: ahh ok i thought there was some special markup to use.. i guess this can be re-word this as a numbered list then

> ### Configuration and credentials

Reviewer: @aleks-mariusz Don't worry, I'll fix this.

> ### Configuration and credentials

Reviewer: @nathan-px @aleks-mariusz It's fixed.

      rabbitmq_federation,
      rabbitmq_federation_management,
      rabbitmq_consistent_hash_exchange,
      rabbitmq_management,
      rabbitmq_peer_discovery_k8s
    ].
  rabbitmq.conf: |
    ## RabbitMQ configuration
    ## Ref: https://github.com/rabbitmq/rabbitmq-server/blob/master/docs/rabbitmq.conf.example

    ## Authentification

    ## Clustering
    cluster_formation.peer_discovery_backend  = rabbit_peer_discovery_k8s
    cluster_formation.k8s.host = kubernetes.default.svc.cluster.local
    cluster_formation.k8s.address_type = hostname
    cluster_formation.node_cleanup.interval = 10
    # Set to false if automatic cleanup of absent nodes is desired.
    # This can be dangerous, see http://www.rabbitmq.com/cluster-formation.html#node-health-checks-and-cleanup.
    cluster_formation.node_cleanup.only_log_warning = true
    cluster_partition_handling = autoheal

    ## The default "guest" user is only permitted to access the server
    ## via a loopback interface (e.g. localhost)
    loopback_users.guest = false

    management.load_definitions = /etc/definitions/definitions.json

    ## Memory-based Flow Control threshold
    vm_memory_high_watermark.absolute = 256MB

>     ## Auth HTTP Backend Plugin

Reviewer 2: Is upstream within Portworx? or elsewhere? Maybe we can get clarification from the SME.

>     ## Auth HTTP Backend Plugin

Reviewer: @nathan-px I think Helm automatically generated these comments and @aleks-mariusz just copied and pasted them here. Overall I find them useful but maybe we can drop the second hash mark.

>     ## Auth HTTP Backend Plugin

Author: andrei is correct, the upstream source _are_ the helm templates themselves [configmap file](https://github.com/helm/charts/blob/master/stable/rabbitmq-ha/templates/configmap.yaml)

            - weight: 1
              podAffinityTerm:
                topologyKey: "kubernetes.io/hostname"
                labelSelector:
                  matchLabels:
                    app: rabbitmq-ha
      volumes:
        - name: config
          emptyDir: {}
        - name: configmap
          configMap:
            name: rmq-rabbitmq-ha
        - name: definitions
          secret:
            secretName: rmq-rabbitmq-ha
            items:
            - key: definitions.json
              path: definitions.json
  volumeClaimTemplates:
    - metadata:
        name: data
        annotations:
      spec:
        accessModes:
          - "ReadWriteOnce"
        resources:
          requests:
            storage: "8Gi"
        storageClassName: "portworx-rabbitmq"
_EOF
```

```output
service/rmq-rabbitmq-ha-discovery created
service/rmq-rabbitmq-ha created
statefulset.apps/rmq-rabbitmq-ha created
```

Note that the last line of the spec references the `portworx-rabbitmq` storage class defined in the [Create a StorageClass](#create-a-storage-class) section. As a result, Kubernetes will automatically create a new PVC for each replica.

## Validate the cluster functionality

> Regardless of the method you used to set up RabbitMQ, you can test your installation using a tool called [PerfTest](https://rabbitmq.github.io/rabbitmq-perf-test/stable/htmlsingle/). In the next sections, we will use PerfTest to:

Author: fixed in 56fe6db

* simulate basic workloads
* measure the performance of the system
* introduce a failure and see how the cluster recovers from it.

### Example RabbitMQ policy for H/A

> First, we configure a policy named `perf-test-with-ha`  with the following settings:

Author: fixed in ec66ebb

> * It's only applied on queues that begin with `perf-test` <!--revisit this -->

Reviewer: @nathan-px Yeah, I wanted to revisit this. Thanks!

* Sets the [ha-mode parameter].(https://www.rabbitmq.com/ha.html#mirroring-arguments) equal to `2`.
* Places the [queue-masters](https://www.rabbitmq.com/ha.html#behaviour) on the least-loaded node.
* Sets up the queues as [lazy](https://www.rabbitmq.com/lazy-queues.html). RabbitMQ will save the queues to disk as early as possible.

```text
kubectl exec rmq-rabbitmq-ha-0 -- \
  rabbitmqctl set_policy perf-test-with-ha '^perf-test' \
'{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}' --apply-to queues
```

```output
Setting policy "perf-test-with-ha" for pattern "^perf-test" to "{
  "ha-mode":"exactly",
  "ha-params":2,
  "ha-sync-mode":"automatic",
  "queue-master-locator":"min-masters",
  "queue-mode":"lazy"
}" with priority "0" for vhost "/" ...
```

### Monitoring

RabbitMQ provides a web-based [user interface](https://www.rabbitmq.com/management.html) that you can use to manage your cluster. To access it, you must forward connections made to a local port to the 15672 port on the   `rmq-rabbitmq-ha-0` pod.

> Enter the following `kubectl port-forward` command to forward all connections made to `localhost:15672` to `rmq-rabbitmq-ha-0:15672`:

Reviewer: @nathan-px Now I can make changes to this branch. I'll format this section.

> Enter the following `kubectl port-forward` command to forward all connections made to `localhost:15672` to `rmq-rabbitmq-ha-0:15672`:

Reviewer: @nathan-px @aleks-mariusz Done.

```text
kubectl port-forward rmq-rabbitmq-ha-0 15672:15672
```

```output
Forwarding from 127.0.0.1:15672 -> 15672
```

At this point, you can access the RabbitMQ user interface by pointing your browser to [localhost:1567](localhost:1567). Use the credentials created in the [#set-up-rabbitmq-using-helm].

> Once you start the performance tests, you can then visit the [Queue section](http://127.0.0.1:15672/#/queues), to see that the `perf-test-with-ha` policy is applied.

Author: fixed in 6fa99bc

Additionally, you can stream to `stdout` the logs of the pod that survives the failover test by entering the following command:

```text
kubectl logs -f rmq-rabbitmq-ha-0
```

### Containerized testing environment

Here we spawn a pod with one container which will allow us to run PerfTest in the [_next_ step](#simulate-broker-usage):

```text
kubectl run perftest \
  --restart=Always \
  --image dummy-required-param \
  --generator=run-pod/v1 \
  --overrides \
'{
  "spec": {
    "affinity": {
      "podAntiAffinity": {
        "requiredDuringSchedulingIgnoredDuringExecution": [{
          "labelSelector": {
            "matchExpressions": [{
              "key": "app",
              "operator": "In",
              "values": ["rabbitmq-ha"]
            }]
          },
          "topologyKey": "kubernetes.io/hostname"
        }]
      }
    },
    "containers": [{
      "command": ["sh"],
      "image": "pivotalrabbitmq/perf-test:latest",
      "name": "perftest",
      "args": ["-c", "sleep 3600"]
    }],
    "securityContext": {
      "runAsUser": 1000
    }
  }
}'
```

```output
pod/perftest created
```

{{<info>}}

> Note the above command starts up a testing pod with one sleeping container, to be scheduled onto one of the _other_ nodes (that _aren't_ running RabbitMQ), so as utilze the network equally between both RabbitMQ nodes. 

Reviewer: ```suggestion
Note that the above command creates a pod with one sleeping container. To balance the network traffic amongst the RabbitMQ nodes, we schedule this container on a node that doesn't run RabbitMQ. Also, note that we run these commands as a non-root user. This is useful for when you have [`PodSecurityPolicies`](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) enabled.
```

Author: fixed

Also note that we use a non-root user to run the commands. This is useful for when you have [`PodSecurityPolicies`](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) enabled.
{{</info>}}

In the next step we will `kubectl exec` into this pod.

### Simulate basic workloads

To start the simulated producers and consumers, run the `bin/runjava` command in the `perftest` container:

```text
kubectl exec -it perftest -- \
    bin/runjava com.rabbitmq.perf.PerfTest \
      --time 900 \
      --queue-pattern 'perf-test-%d' \
      --queue-pattern-from 1 \
      --queue-pattern-to 2 \
      --producers 2 \
      --consumers 8 \
      --queue-args x-cancel-on-ha-failover=true \
      --flag persistent \
      --uri amqp://admin:secretpassword@rmq-rabbitmq-ha:5672?failover=failover_exchange
```

There are several things to note about this example:

> * run the session for 15 minutes with queues named in the format of `perf-test-` followed by a number (in our case `1` and `2`)

Author: fixed in 2639e8f

* create threads for 2 producers and 8 consumers (so theoretically each of the two rabbitmq replicas should host 1 queue and get 1 producers and 4 consumers
* set messages to be flagged as persistent (so will always involve IO which should be understood that this trades off performance for reliabilty)
* use the queue with the sample credentials, at the expected address from the previously defined service (and is internal to the kubernetes cluster)

### Failover

> Here we can create a scenario for failure, such as killing one of the pods.

Reviewer: ```suggestion
Simulate a failure by force deleting the `rmq-rabbitmq-ha-1` pod:
```

Author: fixed

pod "rmq-rabbitmq-ha-1" force deleted
```

As the system recovers from the failure, you will see that the performance testing session will pause for a short period. As a result, the messages will pile up in the queue. After the failover, the cluster resumes normal processing, and no messages are lost. 

## Cleaning up

If you used Helm to install the RabbitMQ cluster, then enter the following command to remove all the Kubernetes components and delete the release:

```text
helm delete --purge rmq
```

If you installed the RabbitMQ cluster manually, then enter the following command to delete all the associated resources:

```text
kubectl delete sts/rmq-rabbitmq-ha \
  svc/rmq-rabbitmq-ha svc/rmq-rabbitmq-ha-discovery \
  rolebinding/rmq-rabbitmq-ha \
  role/rmq-rabbitmq-ha \
  sa/rmq-rabbitmq-ha \
  secret/rmq-rabbitmq-ha \
  cm/rmq-rabbitmq-ha
```

Finally, regardless of the deployment method, execute the following command:

```text
kubectl delete pvc/data-rmq-rabbitmq-ha-0 pvc/data-rmq-rabbitmq-ha-1 \
  sc/portworx-rabbitmq \
  po/perftest
```

> The `kubectl delete` commands remove the following:

Author: fixed in 7861461

 * The workload itself (StatefulSet and the associated services)
 * The RBAC settings for the workload (Rolebinding, Role, and ServiceAccount)
 * The RBAC configuration (ConfigMap and secret)
 * The volumes including data (PersistentVolumeClaims)
 * The volume parameters (StorageClass)
 * The pod we used for testing

## Summary

Software as critical to distributed systems as RabbitMQ is to many of its users _needs_ to be set up in a reliable way.  Portworx is a key ingredient to achieve that goal, as demonstrated in this document.

{{<info>}}

> It should be understood that regardless of which storage solution is used, the goal of reliability  trades off some performance, by necessity of disk-IO being in the critical path of message processing.  While for highest performance messages can be handled entirely inside memory (which admittedly is orders of magnitudes faster), using that approach reliability would be partially sacrificed and limiting as memory is a far more finite resource than disk.

Author: using portworx is not impacting performance.. basically it's setting up HA that impacts performance as it complicates the message path.. the same issue would happen given any storage solution

> It should be understood that regardless of which storage solution is used, the goal of reliability  trades off some performance, by necessity of disk-IO being in the critical path of message processing.  While for highest performance messages can be handled entirely inside memory (which admittedly is orders of magnitudes faster), using that approach reliability would be partially sacrificed and limiting as memory is a far more finite resource than disk.

Reviewer: @nathan-px @aleks-mariusz I commented this out until we come up with a better wording.

My 2c: I think we should remove this section.

> It should be understood that regardless of which storage solution is used, the goal of reliability  trades off some performance, by necessity of disk-IO being in the critical path of message processing.  While for highest performance messages can be handled entirely inside memory (which admittedly is orders of magnitudes faster), using that approach reliability would be partially sacrificed and limiting as memory is a far more finite resource than disk.

Reviewer 2: So this is a function of using RabbitMQ with mirrored queues? I agree with @andreipope that we should probably just remove it. 

If we do need to note that running RabbitMQ in HA mode impacts performance, we should add it immediately following the mention of it in the intro concept information.

---

> title: RabbitMQ on Kubernetes on Portworx

Reviewer: RabbitMQ on Kubernetes with Portworx

> title: RabbitMQ on Kubernetes on Portworx

Reviewer: ```suggestion
RabbitMQ with Portworx on Kubernetes
```

Author: fixed

linkTitle: RabbitMQ
keywords: portworx, container, Kubernetes, rabbitMQ, storage, Docker, k8s, pvc

> description: See how stateful RabbitMQ can be deployed on Kubernetes using Portworx volumes.

Reviewer: ```suggestion
description: See how RabbitMQ can be deployed on Kubernetes using Portworx volumes.
```

Author: fixed

weight: 2
noicon: true
---

> [RabbitMQ](https://www.rabbitmq.com/) is the most widely used open-source message-queue/broker software today.  It plays a central role in many distributed systems and can be configured so that messages between decoupled systems are passed and safely persisted, as well as replicated to other nodes.

Reviewer: ```suggestion
[RabbitMQ](https://www.rabbitmq.com/) is an open source message broker which plays a central role in distributed systems. You can configure RabbitMQ to safely pass message between decoupled systems.
```

> [RabbitMQ](https://www.rabbitmq.com/) is the most widely used open-source message-queue/broker software today.  It plays a central role in many distributed systems and can be configured so that messages between decoupled systems are passed and safely persisted, as well as replicated to other nodes.

Reviewer: ```suggestion
[RabbitMQ](https://www.rabbitmq.com/) is an open-source message broker. It plays a central role in distributed systems. You can configure RabbitMQ to safely pass messages between decoupled systems.
```

Author: fixed

> In this reference architecture document, we will explore setting up RabbitMQ this way in the "default" namespace, and leverage these features using Portworx for providing a reliable persistent storage layer, helping make sure no messages are lost.

Reviewer: ```suggestion
This reference architecture document shows how you can create and run a RabbitMQ with Portworx on Kubernetes. This way, Portworx will provide a reliable persistent storage layer, which makes sure no messages are lost.
```

Author: fixed

> Specifically, our RabbitMQ cluster will use mirrored-queues, so that messages data and metadata is persisted to Portworx volumes, which also are replicated, to prevent messages from being lost.  This provides multiple layers of redundancy.

Reviewer: ```suggestion
The RabbitMQ cluster will use mirrored-queues to persist data and metadata to Portworx volumes. Portworx also replicates the volumes, providing multiple layers of redundancy.
```

Author: fixed

> 

Reviewer: ```suggestion

## Prerequisites

* You must have a Kubernetes cluster with a minimum of 3 worker nodes.
* Portworx is installed on your Kubernetes cluster. For more details on how to install Portworx, refer to the instructions from the [Portworx on Kubernetes](/portworx-install-with-kubernetes/) page.
* (Optional) Helm in installed on your Kubernetes cluster. Refer to the [Helm Quickstart Guide](https://helm.sh/docs/using_helm/#quickstart) for details about how to install Helm.

```

Author: added

> ## Portworx-powered volume provisioning

Reviewer: @aleks-mariusz Do the users always need to do this first, no matter how they choose to install RabbitMQ?

Reviewer: Also, I'd suggest "Create a StorageClass"

Author: they need to set the policy described below at least once for each fresh cluster, otherwise a fresh cluster does not have the settings the policy requires to perform queue-mirroring

Reviewer: I see. Thanks!

Reviewer: ![RabbitMQ](https://user-images.githubusercontent.com/32929697/68603653-6648f000-04b1-11ea-851d-0231c3c88f8b.png)
It looks something like this, right?

Author: yes, exactly that's correct (tho you spelled Storag(e)Class wrong in the first block).

Reviewer: @aleks-mariusz Great, thanks!

> RabbitMQ will first need a [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) definition that sets the Portworx storage parameters for volume-creation, which are later attached to the pods that will be created.

Reviewer: Not sure we need this.

Author: i thought a storage class however _is_ needed (and recommended to be customized on an app-by-app basis)

Reviewer: For sure. But do we need to explain what a StorageClass is?

Reviewer: @nathan-px Thoughts?

> This _StorageClass_ will be referenced by a [PersistentVolumeClaims](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) created later in this document.  The PersistentVolume created by that is then used by the RabbitMQ cluster pods.

Reviewer: Not sure we need this.

Author: if it's to use storage, it needs a storageclass, unless i'm misunderstanding what you're referring to that should be deleted?

Reviewer: Agree :) It's just that I'm not sure we need to explain this.

Author: Good point, if we expect the user is familiar enough with Kubernetes, then i'd say no.. However, if we're anticipating the user reading the doc may not be that fluent with Kubernetes, i don't see the the harm in giving them a link to explain the role of what storageclasses do.  If this is an invalid assumption though, i'll remove the perhaps redundant explaination of storageclasses..

Reviewer: @nathan-px Thoughts?

Author: basically, my thinking was that sometimes people will google for rabbitmq and kubernetes, and we want them to of course, not only be able to find our page but ALSO not be put off if they are not super-familiar with kubernetes concepts like storageclasses..

Reviewer: @nathan-px ditto.

> The following will specifying each volume should have two replicas (of the data), and that it be backed by a high-IO priority storage pool (which internally to Portworx, is a collection of similar spec disks/devices).

Reviewer: ```suggestion
The spec below creates a `StorageClass` with 3 replicas and uses a high-priority storage pool:
```

Author: fixed

> Run:

Reviewer: Delete.

Author: done

> ```

Reviewer: ```suggestion
```text
```sugggestion

> ```

Reviewer: ```text

> ```

Reviewer: Should be "```text```"

> ```

Reviewer: text

Author: added

kubectl apply -f - <<'_EOF'
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: portworx-rabbitmq
parameters:
  io_priority: high
  repl: "2"
  group: "rmq_vg"
provisioner: kubernetes.io/portworx-volume
allowVolumeExpansion: true
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
_EOF
```

> ..to which you should get back the following response:

Reviewer: delete

Author: not deleting yet until we're on the same page about including a storageclass in a ref-arch (as other ones have).

Author: actually, the `to which... following response` part has been deleted elsewhere

> ```

Reviewer: output

Author: added

> ...

Reviewer: delete

Author: removed

storageclass.storage.k8s.io/portworx-rabbitmq created
```

> For details on what all the above parameters do, please consult the the relevant Kubernetes [storageclass documentation](https://kubernetes.io/docs/concepts/storage/storage-classes/#the-storageclass-resource) or the [Portworx-specific section](https://kubernetes.io/docs/concepts/storage/storage-classes/#portworx-volume) of that documentation.

Reviewer: ```suggestion
For details about the Portworx-specific parameters, refer to the [Powrtorx Volume]https://kubernetes.io/docs/concepts/storage/storage-classes/#portworx-volume] section.
```

Author: fixed

{{<info>}}

> The above assumes your Kubernetes is _not_ setup to use [CSI](https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/). If you are then instead set the `provisioner` parameter to `pxd.portworx.com`

Reviewer: ```suggestion
If you're using Portworx with CSI, you need to set the value of the `provisioner` parameter to `pxd.portworx.com`.
```

Author: fixed

{{</info>}}

> 

Reviewer: ```suggestion
When you install RabbitMQ with Portworx on Kubernetes, you can choose one of the following options:

* Set up RabbitMQ using [Helm](https://helm.sh)]
* Set up RabbitMQ manually
```

Author: added

> ## Setup RabbitMQ (using Helm)

Reviewer: ```suggestion
Set up RabbitMQ using Helm
```

Author: indentation fixed

### Helm

> Applications running on Kubernetes typically have _several_ yaml files defining the differnet components.  The more complex an app, the more of these one has to deal with.

Reviewer: ```suggestion
A complex application like RabbitMQ relies on several YAML files to define its various components. The next chapter will show how you can use Helm to simplify the deployment of RabbitMQ.
```

Author: fixed

In order to simply a deployment of a system as complex as RabbitMQ, we will utilize [Helm](https://helm.sh) to simply deployment.

If you do not have Helm set up, you can either do so by consulting [their documentation](https://helm.sh/docs/using_helm/#quickstart), or skip ahead to the [next section](#setup-rabbitmq-manually) where we install RabbitMQ in the classic, more involved way (without using Helm).

### Launching RabbitMQ release

The following Helm command will use the [rabbitmq-ha chart](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha) to create (or update) a [release](https://github.com/helm/helm/blob/release-2.14/docs/glossary.md#release) "instance" named `rmq`, which will form a two-member RabbitMQ cluster.  This allows queues and messages to be mirrored between them.

> Run:

Reviewer: This should be removed.

Author: done

> ```

Reviewer: text

Author: added
